pkg/storage/utils/decomposedfs/spaces.go

```diff
@@ -300,28 +303,75 @@ func (fs *Decomposedfs) ListStorageSpaces(ctx context.Context, filter []*provide
 
 	if requestedUserID != nil {
 		allMatches := map[string]string{}
-		indexPath := filepath.Join(fs.o.Root, "indexes", "by-user-id", requestedUserID.GetOpaqueId())
+		indexPath := filepath.Join(fs.o.Root, "indexes", "by-user-id", requestedUserID.GetOpaqueId()+".mpk")
 		fi, err := os.Stat(indexPath)
 		if err == nil {
 			allMatches, err = fs.spaceIDCache.LoadOrStore("by-user-id:"+requestedUserID.GetOpaqueId(), fi.ModTime(), func() (map[string]string, error) {
-				path := filepath.Join(fs.o.Root, "indexes", "by-user-id", requestedUserID.GetOpaqueId(), "*")
-				m, err := filepath.Glob(path)
+				// Acquire a read log on the index file
+				f, err := lockedfile.Open(indexPath)
 				if err != nil {
-					return nil, err
+					return nil, errors.Wrap(err, "unable to lock index to read")
 				}
-				matches := map[string]string{}
-				for _, match := range m {
-					link, err := os.Readlink(match)
+				defer func() {
+					rerr := f.Close()
+
+					// if err is non nil we do not overwrite that
+					if err == nil {
+						err = rerr
+					}
+				}()
+
+				// Read current state
+				msgBytes, err := io.ReadAll(f)
+				if err != nil {
+					return nil, errors.Wrap(err, "unable to read index")
+				}
+				links := map[string]string{}
+				if len(msgBytes) > 0 {
+					err = msgpack.Unmarshal(msgBytes, &links)
 					if err != nil {
-						continue
+						return nil, errors.Wrap(err, "unable to parse index")
 					}
-					matches[match] = link
 				}
-				return matches, nil
+				return links, nil
 			})
-		}
-		if err != nil {
-			return nil, err
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			indexPath := filepath.Join(fs.o.Root, "indexes", "by-user-id", requestedUserID.GetOpaqueId())
```

Comment:
This migration code could also go into a decomposedfs migration which would make this code leaner and cleaner.
That would also make sure that `UserIDToUserAndGroups` can always read from messagepack.